### PR TITLE
feat(BlockPlugin): block plugin support of GetHeaderByHash

### PR DESCRIPTION
**Pull request by archbear**
This PR fixes Block/Historical Plugin code paths so that:
- all queries for header go through block plugin
- all queries for block go through historical plugin

### CODE REVIEW

**Comment by itsdevbear** [comment]
You don't wanna store anything by hash in the IAVL tree since it will cause state growth.

**Comment by itsdevbear** [comment]
If we really wanted we could do a 256 block revolving queue, but idk if thats the move. Whats the issue that this PR is trying to solve?

**Comment by archbear** [comment]
> If we really wanted we could do a 256 block revolving queue, but idk if thats the move. Whats the issue that this PR is trying to solve?

The get genesis by header and by number rpc queries were causing the backend to call the historical plugin instead of the block plugin. That lead to trying to distinguish the responsibilities of the block and historical plugin from header to txs etc.

File: eth/polar/backend.go [hunk 1]
```diff
@@ -218,21 +218,22 @@ func (b *backend) HeaderByNumberOrHash(ctx context.Context,
...
-	return b.polar.blockchain.GetHeaderByHash(hash), nil
+	header := b.polar.blockchain.GetHeaderByHash(hash)
+	if header == nil {
+		return nil, errors.New("header for hash not found")
+	}
+
+	// if blockNrOrHash.RequireCanonical && b.eth.blockchain.GetCanonicalHash(header.Number.Uint64()) != hash {
+	// 	return nil, errors.New("hash is not currently canonical")
+	// }
+	return header, nil
```
Comment (calbera) [inline eth/polar/backend.go:236]:
```suggestion
	return b.polar.blockchain.GetHeaderByHash(hash)
```

File: eth/core/chain_reader.go [hunk 2]
```diff
@@ -116,11 +116,11 @@ func (bc *blockchain) CurrentSafeBlock() *types.Header {
...
 // found.
 func (bc *blockchain) GetHeaderByHash(hash common.Hash) *types.Header {
-	block := bc.GetBlockByHash(hash)
-	if block == nil {
+	header, err := bc.bp.GetHeaderByHash(hash)
+	if header == nil || err != nil {
 		return nil
 	}
-	return block.Header()
+	return header
```
Comment (calbera) [inline eth/core/chain_reader.go:123]: `return bc.bp.GetHeaderByHash(hash)`

File: cosmos/x/evm/types/keys.go [hunk 3]
```diff
@@ -29,8 +29,10 @@ const (
 	BlockHashKeyToReceiptsPrefix
 	TxHashKeyToTxPrefix
 	VersionKey
-	HeaderKey
-	GenesisHeaderKey
+	HeaderNumKey
+	HeaderHashKey
```
Comment (calbera) [inline cosmos/x/evm/types/keys.go:33]:
```suggestion
	HeaderHashKeyPrefix
```

**Review by calbera — APPROVED** [review]
tACK

**Comment by calbera** [comment]
@itsdevbear this is good to go, integration tests are very flaky rn. also broken on main

File: eth/core/chain_reader.go [hunk 4]
```diff
@@ -38,14 +38,14 @@ type ChainReader interface {
 // ChainBlockReader defines methods that are used to read information about blocks in the chain.
 type ChainBlockReader interface {
 	CurrentHeader() *types.Header
-	CurrentBlock() *types.Header
```
Comment (itsdevbear) [inline eth/core/chain_reader.go:41]: this does not match geth.

File: eth/polar/backend.go [hunk 5]
```diff
@@ -398,7 +393,7 @@ func (b *backend) GetLogs(
 // GetTd returns the total difficulty of a block in the canonical chain.
 // This is hardcoded to 69, as it is only applicable in a PoW chain.
 func (b *backend) GetTd(_ context.Context, hash common.Hash) *big.Int {
-	if header := b.polar.blockchain.GetHeaderByHash(hash); header != nil {
+	if header, err := b.polar.blockchain.GetHeaderByHash(hash); err == nil {
```
Comment (itsdevbear) [inline eth/polar/backend.go:396]: doesn't match geth

File: eth/core/chain_reader.go (hunk 4 above)
Comment (calbera) [reply eth/core/chain_reader.go:41]: Same function

File: eth/polar/backend.go (hunk 5 above)
Comment (calbera) [reply eth/polar/backend.go:396]: Just propogating error, useful for the client making call

**Comment by calbera** [comment]
@itsdevbear going to close this branch and the block plugin refactor should be on @archbear’s new branch. But we can revert some of the new variances with geth. Was mainly for passing the error up the stack to the RPC call

File: cosmos/x/evm/plugins/block/header_test.go [hunk 6]
```diff
@@ -39,70 +46,108 @@ var _ = Describe("Header", func() {
...
+		Expect(err).ToNot(HaveOccurred())
+		Expect(header2.Hash()).To(Equal(header.Hash()))
+
+		// get unknown header should return the latest header (10)
+		header3, err := p.GetHeaderByNumber(11)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(header3.Hash()).To(Equal(header.Hash()))
+	})
+
+	FIt("should be able to prune headers", func() {
```
Comment (itsdevbear) [inline cosmos/x/evm/plugins/block/header_test.go:97]: focus

**Review by itsdevbear — COMMENTED** [review]
add test for pruning

**Review by itsdevbear — APPROVED** [review]

**Comment by itsdevbear** [comment]
@calbera why removing hive from ci?

**Comment by calbera** [comment]
@itsdevbear cuz its just stuck like this 
image
re-enable in the other PR

**Review by itsdevbear — APPROVED** [review]

**Pull request merged**